velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,316 @@
+  std::stringstream out;
+  out << "[";
+  std::for_each(
+      input.begin(), input.end(), [&out](const auto& e) { out << e << ", "; });
+  out << "]";
+  return out.str();
+}
+
+template <typename T, bool returnsOptionalValues>
+exec::ArrayView<returnsOptionalValues, T> read(
+    exec::VectorReader<Array<T>>& reader,
+    size_t offset) {
+  if constexpr (returnsOptionalValues) {
+    return reader[offset];
+  } else {
+    return reader.readNullFree(offset);
+  }
+}
+} // namespace
+
+namespace {
+class ArrayShuffleTest : public FunctionBaseTest {
+ protected:
+  constexpr static const int32_t kNumShuffleTimes = 100;
+  constexpr static const int32_t kNums = 10;
+  constexpr static const double kUniquenessRate = .7;
+
+  template <typename T, bool returnsOptionalValues = false>
+  void testShuffle(const VectorPtr& input) {
+    DecodedVector decodedExpected;
+    decodedExpected.decode(*input.get());
+    exec::VectorReader<Array<T>> readerExpected(&decodedExpected);
+
+    auto actualVector =
+        evaluate<ArrayVector>("shuffle(C0)", makeRowVector({input}));
+    // Validate each row from the actual decoded ArrayVector is a permutation
+    // of the corresponding row from the expected decoded ArrayVector.
+    DecodedVector decodedActual;
+    decodedActual.decode(*actualVector.get());
+    exec::VectorReader<Array<T>> readerActual(&decodedActual);
+
+    for (auto i = 0; i < input->size(); i++) {
+      // Must materialize into std::vector, otherwise it will throw error
+      // because ArrayView doesn't support std::is_permutation() yet.
+      auto actualArray =
+          read<T, returnsOptionalValues>(readerActual, i).materialize();
+      auto expectedArray =
+          read<T, returnsOptionalValues>(readerExpected, i).materialize();
+
+      // Assert the two vectors containing the same elements (ignoring order).
+      ASSERT_TRUE(std::is_permutation(
+          actualArray.begin(), actualArray.end(), expectedArray.begin()))
+          << "Actual array " << printArray(actualArray) << " at " << i
+          << " must produce a permutation of expected array "
+          << printArray(expectedArray);
+    }
+  }
+
+  /// Generate a range-based array N {0, 1, ..., n-1} as the input for
+  /// test shuffle randomness purpose.
+  /// 1. For flat encoding: we generate an t-size ArrayVector with identical
+  /// elements (N).
+  /// 2. For constant encoding: we generate an t-size ConstantVector with
+  /// valueVector = N.
+  /// 3. For dict encoding: we generate a DictionaryVector with a 2t-size base
+  /// ArrayVector (with and identical elements: N) and a t-size indices.
+  VectorPtr generateVector(
+      const VectorEncoding::Simple& encoding,
+      const int32_t n = kNums,
+      const int32_t t = kNumShuffleTimes) {
+    std::vector<int32_t> inputData(n);
+    std::iota(inputData.begin(), inputData.end(), 0);
+
+    if (encoding == VectorEncoding::Simple::FLAT) {
+      std::vector<std::vector<int32_t>> flatData(t);
```

Comment:
Using single-letter variables is an anti pattern because it makes code hard to read. Can we use descriptive names for t and n?